src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -206,32 +241,96 @@ dfs vs = concatMap flatten . dfsForestFrom vs
 -- 'isSubgraphOf' ('vertices' $ reachable x y) y == True
 -- @
 reachable :: Int -> AdjacencyIntMap -> [Int]
-reachable x = concat . bfs [x]
+reachable x = dfs [x]
+
+type Cycle = NonEmpty
+type TopSort = Either (Cycle Int) [Int]
+type Entry = Either Int Int
+type ParentTable = IntMap.IntMap Entry
+data S = S { table :: !ParentTable, order :: [Int] }
+
+-- Tree edge when node has not been explored
+pattern TreeEdge :: Maybe Entry
+pattern TreeEdge <- Nothing
+-- Back edge when node is discovered, but not done
+pattern BackEdge :: Maybe Entry
+pattern BackEdge <- Just (Left _)
+-- Pattern to retrieve Parent from table lookup
+pattern Parent :: Int -> Entry
+pattern Parent p <- Left p
+
+topSort' :: (MonadState S m, MonadCont m) => AdjacencyIntMap -> m TopSort
+topSort' g = callCC $ \cyclic ->
+  do let vertices = map fst $ IntMap.toDescList $ adjacencyIntMap g
+         adjacent = IntSet.toDescList . flip postIntSet g
+         dfs z x =
```

Comment:
Shall we give a bit more descriptive names to variables? I suggest: `z -> prev`, `x -> cur`, `y -> next`.

(I'm slowly going through the implementation, so apologies if didn't get the meaning right.)

Reply:
Yes, z was prev, x current :). I can also get rid of the name shadowing if it makes the code too cryptic.

When the representation was more convoluted, I found the PatternSynonyms were kind of necessary to avoid horrible pattern matching. Currently, things are a bit simpler, and I could do away with them?

Reply:
Yes, I think in this case you could rename `classify` to something like `getBackEdge` and then results `Nothing` and `Just _` may become clear without pattern synonyms. I haven't yet figured out what `Left` and `Right` mean, some more comments (perhaps at the top next to the definitions of data types/synonyms) would be helpful!

Reply:
Ah, yes that makes sense, a suitable name should make the patterns clear. 

`enter prev curr` enteres `(Left prev)` as the parent of `curr`, and on exiting promotes it to `Right prev`. I was considering just defining a datatype along the lines of:

`data Entry a = Root | Started a | Finished a` 

to indicate parents. Then entering `undefined` for `prev` could also be avoided when starting a new component

Reply:
Ah, yes, that seems like a nice naming scheme! I hope it doesn't come with any performance penalty?

Reply:
Yeah same, I'll try it and double check

Reply:
> @jitwit Yes, we might do that at some point, but for now, let's just define type synonyms locally. In the case of `Bipartite.OddCycle` we might at some point switch a data type that forces the odd length.
> 
> P.S.: This and the next weeks are quite busy, so apologies for the delays.

Sounds good. Also no worries.

By the way, in thinking about complexity analysis for `dfs` I realized the ones for `bfs` were either for a previous implementation or were just wrong! Should I open a separate pull request to correct them?

Reply:
> Should I open a separate pull request to correct them?

Ideally, unrelated changes should be in a separate PR, but I don't feel too strongly about minor fixes, typos, etc., so I'll let you decide :)